go-controller/pkg/ovn/namespace.go

```diff
@@ -350,18 +350,19 @@ func (oc *Controller) deleteNamespace(ns *kapi.Namespace) {
 	klog.V(5).Infof("Deleting Namespace's NetworkPolicy entities")
 	for _, np := range nsInfo.networkPolicies {
 		key := getPolicyNamespacedName(np.policy)
-		oc.retryNetworkPolicies.skipRetryObj(key)
+		lockedKey := oc.retryNetworkPolicies.lockKey(key)
+		// unlock key at the end of the iteration - can't use defer here
 		// add the full np object to the retry entry, since the namespace is going to be removed
 		// along with any mappings of nsInfo -> network policies
-		oc.retryNetworkPolicies.initRetryObjWithDelete(np.policy, key, np, false)
+		oc.retryNetworkPolicies.initRetryObjWithDelete(np.policy, lockedKey, np, false)
 		isLastPolicyInNamespace := len(nsInfo.networkPolicies) == 1
 		if err := oc.destroyNetworkPolicy(np, isLastPolicyInNamespace); err != nil {
 			klog.Errorf("Failed to delete network policy: %s, error: %v", key, err)
-			oc.retryNetworkPolicies.unSkipRetryObj(key)
 		} else {
-			oc.retryNetworkPolicies.deleteRetryObj(key, true)
+			oc.retryNetworkPolicies.deleteRetryObj(lockedKey)
```

Comment:
Have a `AddRetryObjWithDelete` and `DeleteRetryObject` in retryObjs that handles key locking internally